#### Exercici 1. Crea almenys dos models de regressió diferents per intentar predir el millor possible el preu de les vivendes (MEDV) de l'arxiu adjunt.

NOTA: Partim del dataset que ha estat netejat i transformat en el Sprint 5.

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
housing_ML = pd.read_excel('/Users/juliacorominasfont/Desktop/DATA ANALYST/Bootcamp IT Academy/Tasques Sprints/Sprint_6_ML_Regressions/housing_prepared.xlsx')
housing_ML.head(5)

Unnamed: 0      CRIM    ZN     INDUS       NOX        RM       AGE  \
0           0 -0.069593  1.44 -0.571650  0.000000  0.496612 -0.250765   
1           1 -0.063755  0.00 -0.202943 -0.394286  0.287940  0.028542   
2           2 -0.063760  0.00 -0.202943 -0.394286  1.323171 -0.334353   
3           3 -0.062347  0.00 -0.581720 -0.457143  1.069783 -0.646279   
4           4 -0.052144  0.00 -0.581720 -0.457143  1.271680 -0.475025   

        DIS       TAX   PTRATIO  ...  RAD_1  RAD_2  RAD_24  RAD_3  RAD_4  \
0  0.285777 -0.087855 -1.339286  ...      1      0       0      0      0   
1  0.569789 -0.227390 -0.446429  ...      0      1       0      0      0   
2  0.569789 -0.227390 -0.446429  ...      0      1       0      0      0   
3  0.924391 -0.279070 -0.125000  ...      0      0       0      1      0   
4  0.924391 -0.279070 -0.125000  ...      0      0       0      1      0   

   RAD_5  RAD_6  RAD_7  RAD_8  MEDV  
0      0      0      0      0  24.0  
1      0      0      0      0  21.6  
2      0      0      0      0  34.7  
3      0      0      0      0  33.4  
4      0      0      0      0  36.2  

[5 rows x 23 columns]

In [3]:
#Eliminació de columnes no rellevants
housing_ML.drop('Unnamed: 0', axis=1, inplace=True)
housing_ML.head(5)

CRIM    ZN     INDUS       NOX        RM       AGE       DIS       TAX  \
0 -0.069593  1.44 -0.571650  0.000000  0.496612 -0.250765  0.285777 -0.087855   
1 -0.063755  0.00 -0.202943 -0.394286  0.287940  0.028542  0.569789 -0.227390   
2 -0.063760  0.00 -0.202943 -0.394286  1.323171 -0.334353  0.569789 -0.227390   
3 -0.062347  0.00 -0.581720 -0.457143  1.069783 -0.646279  0.924391 -0.279070   
4 -0.052144  0.00 -0.581720 -0.457143  1.271680 -0.475025  0.924391 -0.279070   

    PTRATIO         B  ...  RAD_1  RAD_2  RAD_24  RAD_3  RAD_4  RAD_5  RAD_6  \
0 -1.339286  0.261902  ...      1      0       0      0      0      0      0   
1 -0.446429  0.261902  ...      0      1       0      0      0      0      0   
2 -0.446429  0.066675  ...      0      1       0      0      0      0      0   
3 -0.125000  0.153016  ...      0      0       0      1      0      0      0   
4 -0.125000  0.261902  ...      0      0       0      1      0      0      0   

   RAD_7  RAD_8  MEDV  
0      0      0  24.0  
1      0      0  21.6  
2      0      0  34.7  
3      0      0  33.4  
4      0      0  36.2  

[5 rows x 22 columns]

In [4]:
housing_ML.shape

(506, 22)

In [5]:
housing_ML.columns

Index(['CRIM', 'ZN', 'INDUS', 'NOX', 'RM', 'AGE', 'DIS', 'TAX', 'PTRATIO', 'B',
       'LSTAT', 'CHAS', 'RAD_1', 'RAD_2', 'RAD_24', 'RAD_3', 'RAD_4', 'RAD_5',
       'RAD_6', 'RAD_7', 'RAD_8', 'MEDV'],
      dtype='object')

El dataset transformat té 22 columnes, amb les variables següents:
- CRIM: Índex de criminalitat per càpita CRIM per ciutat
- ZN: proporció de sòl residencial zonificat per a lots de més de 25.000 peus quadrats
- INDUS: proporció d'acres comercials no minoristes per ciutat
- CHAS: Variable binària fictícia de Charles River (= 1 si la zona està al costat del riu; 0 en cas contrari). 
- NOX: concentració d'òxids nítrics (parts per 10 milions)
- RM: nombre mitjà d'habitacions per habitatge
- AGE: proporció d'unitats ocupades pels seus propietaris construïdes abans de 1940
- DIS: distàncies ponderades a cinc centres d'ocupació de Boston
- RAD_1, RAD_2, RAD_24, RAD_3, RAD_4, RAD_5, RAD_6, RAD_7, RAD_8, aquestes columnes s'han format a partir de la variable RAD, que fa referència a l'índex d'accessibilitat a autopistes radials.
- TAX: taxa de l'impost sobre béns immobles de valor íntegre per 10.000 dòlars
- PTRATIO: ràtio alumne-professor per ciutat
- B: 1000(Bk - 0,63)^2 on Bk és la proporció de negres per ciutat
- LSTAT: % estatus inferior de la població
- MEDV: Valor mitjà dels habitatges ocupats pels seus propietaris en 1000$. MEDV és la variable objectiu o target.

#### (a) Features (X) i target (y). Conversió de les dades en arrays:

In [6]:
from sklearn.model_selection import train_test_split

X = housing_ML.drop('MEDV', axis=1) #característiques(features)
y = np.array(housing_ML['MEDV']) #variable target (la que volem predir)

X_list = list(X.columns) #Guardem el nom de les variables X en una llista per futur ús
X = np.array(X) #convertim el conjunt X en un np array perquè l'algoritme ho pugui llegir

#### (b) Train i Test Set:

In [7]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print('Forma del Training set: X={}, y={}'.format(X_train.shape, y_train.shape))
print('Forma del Test set: X={}, y={}'.format(X_test.shape, y_test.shape))

Forma del Training set: X=(404, 21), y=(404,)
Forma del Test set: X=(102, 21), y=(102,)


### Models

#### Decision Tree Regression

In [13]:
from sklearn.tree import DecisionTreeRegressor

dt = DecisionTreeRegressor(random_state=42)
dt.fit(X_train, y_train)

DecisionTreeRegressor(random_state=42)

In [14]:
predictions_dt = dt.predict(X_test)

#### Random forest

In [15]:
from sklearn.ensemble import RandomForestRegressor

rf = RandomForestRegressor(random_state = 42)
rf.fit(X_train, y_train)

RandomForestRegressor(random_state=42)

In [16]:
predictions_rf = rf.predict(X_test)

#### Exercici 2. Compara’ls en base al MSE (Mean Square Error) i al R2.

- MSE (error quadràtic mitjà): És una funció de risc que mesura lo aprop que està una línia de regressió d'un conjunt de punts de dades. Ajuda a quantificar la precisió del model. És convenient utilitzar aquesta mesura si la variable objectiu està normalment distribuïda i es vol que els errors grans es penalitzin més que els petits.

- R2 (coeficient de determinació): La puntuació R2 indica si el model és capaç d'explicar bé la variabilitat de la variable objectiu. Cal tenir en compte que si el R2 és molt alt pot ser indicatiu d'un sobreajustament, si no es compleixen altres assumpcions del model.

In [17]:
from sklearn.metrics import mean_squared_error
print('MSE del Decision Tree Regression:', mean_squared_error(y_test, predictions_dt))
print('MSE del Random Forest:', mean_squared_error(y_test, predictions_rf))

MSE del Decision Tree Regression: 11.331568627450983
MSE del Random Forest: 8.012439598039217


El model Random Forest té un MSE més petit que el Decision Tree Regression. Això ens indica que l'error del primer és menor i, per tant, que té un millor rendiment predictiu. Dit d'una altra manera, el model Random Forest és més acurat a l'hora de fer les prediccions.

Cal tenir en compte que el MSE es mesura en les unitats quadrades de la variable objectiu, és a dir, per obtenir una mesura en les mateixes unitats que la variable objectiu es podria utilitzar l'arrel quadrada de l'MSE.

In [18]:
from sklearn.metrics import r2_score
print('R2 del Decision Tree Regression:', r2_score(y_test, predictions_dt))
print('R2 del Random Forest:', r2_score(y_test, predictions_rf))

R2 del Decision Tree Regression: 0.8454795574547436
R2 del Random Forest: 0.8907401302272603


Els dos models tenen un R2 pròxim a 1, el que vol dir que els dos semblen funcionar bé a l'hora d'explicar la variabilitat de la variable objectiu (MEDV). Els resultats indiquen que el model Decision Tree Regression explica el voltant del 84,55% de la variabilitat de la variable target, mentre que el model Random Forest és capaç d'explicar el 89,07%. Per tant, el segon és més capaç d'identificar patrons en el conjunt de dades.

#### Exercici 3. Entrena’ls utilitzant els diferents paràmetres que admeten per intentar millorar-ne la predicció.

In [20]:
from sklearn.model_selection import GridSearchCV

# Definició de la cuadrícula de paràmetres pel DecisionTreeRegressor
param_grid_dt = {
    'max_depth': [3, 5, 7, 10, None],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4, 8],
    'random_state': [42]
}

# Definició de la cuadrícula de paràmetres pel RandomForestRegressor
param_grid_rf = {
    'n_estimators': [50, 100, 200],
    'max_depth': [3, 5, 7, 10, None],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4, 8],
    'random_state': [42]
}

dt_regressor = DecisionTreeRegressor()
rf_regressor = RandomForestRegressor()

# Ús del GridSearchCV per trobar els millors paràmetres pel DecisionTreeRegressor
grid_search_dt = GridSearchCV(dt_regressor, param_grid_dt, cv=5, scoring='neg_mean_squared_error')
grid_search_dt.fit(X_train, y_train)

# Ús del GridSearchCV per trobar els millors paràmetres pel RandomForestRegressor
grid_search_rf = GridSearchCV(rf_regressor, param_grid_rf, cv=5, scoring='neg_mean_squared_error')
grid_search_rf.fit(X_train, y_train)

print("Millors paràmetres pel DecisionTreeRegressor:", grid_search_dt.best_params_)
print("Millors paràmetres pel RandomForestRegressor:", grid_search_rf.best_params_)

Millors paràmetres pel DecisionTreeRegressor: {'max_depth': 7, 'min_samples_leaf': 8, 'min_samples_split': 2, 'random_state': 42}
Millors paràmetres pel RandomForestRegressor: {'max_depth': 10, 'min_samples_leaf': 2, 'min_samples_split': 2, 'n_estimators': 100, 'random_state': 42}


Els paràmetres òptims serien:

- Pel DecisionTreeRegressor:
    - max_depth: 7
    - min_samples_leaf (número mínim de mostres per estar en un node d'una fulla): 8
    - min_samples_split (número mínim de mostres per dividir un node intern): 2

- Pel RandomForestRegressor:
    - max_depth: 10
    - min_samples_leaf (número mínim de mostres per estar en un node d'una fulla): 2
    - min_samples_split (número mínim de mostres per dividir un node intern): 2
    - n_estimators (número d'arbres): 100 

In [21]:
# Prediccions amb els nous models
predictions_best_dt = grid_search_dt.predict(X_test)
predictions_best_rf = grid_search_rf.predict(X_test)

# Evaluació dels models
print("\nMSE del nou DecisionTreeRegressor:", mean_squared_error(y_test, predictions_best_dt))
print("MSE del nou RandomForestRegressor:", mean_squared_error(y_test, predictions_best_rf))

print("\nMillor R2 pel DecisionTreeRegressor:", grid_search_dt.best_score_)
print("Millor R2 pel RandomForestRegressor:", grid_search_rf.best_score_)


MSE del nou DecisionTreeRegressor: 15.92129611952102
MSE del nou RandomForestRegressor: 9.111811406525728

Millor R2 pel DecisionTreeRegressor: -22.31321410542953
Millor R2 pel RandomForestRegressor: -14.827677545579206


Veiem que el rendiment dels models ha empitjorat enlloc de millorar. Això pot ser degut a varies raons, com el overfitting (que els models hagin estat massa ajustats als detalls específics del conjunt d'entrenament i, per tant, no generalitzin tan bé a noves dades) o el fet que els paràmatres no siguin els òptims. Una manera de solucionar-ho és reduïnt la complexitat dels models.

In [26]:
# Nous paràmatres per reduïr la complexitat del model

# Definició de la cuadrícula dels nous paràmetres pel DecisionTreeRegressor
param_grid_dt_simplified = {
    'max_depth': [3, 5, None], #reduïm la profunditat màxima de l'arbre
    'min_samples_split': [2, 5],
    'min_samples_leaf': [1, 2, 4],
    'random_state': [42]
}

# Definició de la cuadrícula dels nous paràmetres pel RandomForestRegressor
param_grid_rf_simplified = {
    'n_estimators': [50, 100], #reduïm el número d'estimadors
    'max_depth': [3, 5, None],
    'min_samples_split': [2, 5],
    'min_samples_leaf': [1, 2, 4],
    'random_state': [42]
}

# Ús del GridSearchCV per trobar els millors paràmetres pel DecisionTreeRegressor
grid_search_dt_simplified = GridSearchCV(dt_regressor, param_grid_dt_simplified, cv=5, scoring='neg_mean_squared_error')
grid_search_dt_simplified.fit(X_train, y_train)

# Ús del GridSearchCV per trobar els millors paràmetres pel RandomForestRegressor
grid_search_rf_simplified = GridSearchCV(rf_regressor, param_grid_rf_simplified, cv=5, scoring='neg_mean_squared_error')
grid_search_rf_simplified.fit(X_train, y_train)

print("Millors paràmetres pel DecisionTreeRegressor:", grid_search_dt_simplified.best_params_)
print("Millors paràmetres pel RandomForestRegressor:", grid_search_rf_simplified.best_params_)

Millors paràmetres pel DecisionTreeRegressor: {'max_depth': 5, 'min_samples_leaf': 2, 'min_samples_split': 2, 'random_state': 42}
Millors paràmetres pel RandomForestRegressor: {'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 100, 'random_state': 42}


In [27]:
# Prediccions amb els nous models
predictions_best_dt_simplified = grid_search_dt_simplified.predict(X_test)
predictions_best_rf_simplified = grid_search_rf_simplified.predict(X_test)

# Evaluació dels models
print("\nMSE del nou DecisionTreeRegressor:", mean_squared_error(y_test, predictions_best_dt_simplified))
print("MSE del nou RandomForestRegressor:", mean_squared_error(y_test, predictions_best_rf_simplified))

print("\nMillor R2 pel DecisionTreeRegressor:", grid_search_dt_simplified.best_score_)
print("Millor R2 pel RandomForestRegressor:", grid_search_rf_simplified.best_score_)


MSE del nou DecisionTreeRegressor: 9.343752254581478
MSE del nou RandomForestRegressor: 8.012439598039217

Millor R2 pel DecisionTreeRegressor: -26.695256574896923
Millor R2 pel RandomForestRegressor: -14.893381867932101


Veiem que, respecte del models inicials (Exercici 2), el rendiment segons el MSE ha millorat. En el cas del DecisionTreeRegressor ha passat de 11.331568627450983 a 9.343752254581478, mentre que en el cas del RandomForestRegressor segueix essent el mateix (8.012439598039217). Això vol que en aquest últim els paràmetres del model simplicat coincideixen amb els del model inicial. Pel que fa al R2, aquesta mètrica ha empitjorat per ambdós models.

Així doncs, es podria seguir ajustar els valors dels paràmetres segons les necessitats del cas concret i continuar iterant fins a trobar el conjunt de paràmetres que millori el rendiment de cada model.

#### Exercici 4. Compara el seu rendiment emprant l’aproximació traint/test o emprant totes les dades (validació interna).

In [30]:
from sklearn.model_selection import cross_val_score
from sklearn.metrics import make_scorer

scorer_r2 = make_scorer(r2_score)

# DecisionTreeRegressor
dt_scores = cross_val_score(dt, X, y, cv=5, scoring='neg_mean_squared_error')
dt_mse_scores = -dt_scores
dt_r2_scores = cross_val_score(dt, X, y, cv=5, scoring=scorer_r2)

# Random Forest Regressor
rf_scores = cross_val_score(rf, X, y, cv=5, scoring='neg_mean_squared_error')
rf_mse_scores = -rf_scores 
rf_r2_scores = cross_val_score(rf, X, y, cv=5, scoring=scorer_r2)

print("Error Quadràtic Mitjà del Decision Tree Regression:", np.mean(dt_mse_scores))
print("R2 del Decision Tree Regression:", np.mean(dt_r2_scores))

print("\nError Quadràtic Mitjà del Random Forest:", np.mean(rf_mse_scores))
print("R2 del Random Forest:", np.mean(rf_r2_scores))

Error Quadràtic Mitjà del Decision Tree Regression: 43.72368452727626
R2 del Decision Tree Regression: 0.06935095041243766

Error Quadràtic Mitjà del Random Forest: 21.270251010774597
R2 del Random Forest: 0.6309561246173885


Els resultats obtinguts amb l'aproximació train/test eren:
- Decision Tree Regressor:
    - MSE: 11.331568627450983
    - R2: 0.8454795574547436
- Random Forest Regressor: 
    - MSE: 8.012439598039217
    - R2: 0.8907401302272603

Veiem que els resultats amb validació creuada mostren un MSE més alt i un R2 més baix que els obtinguts amb l'aproximació train/test, amb una diferència molt més accentuada en el cas del Decision Tree Regressor.

Això pot voler dir que els models no generalitzen tan bé quan s'entrenen i es validen amb les diferents particions del conjunt de dades, ja que obtenen un major error i tenen menys capacitat d'explicar la variabilitat de la variable objectiu. 

#### Exercici 5. No facis servir la variable del nombre d'habitacions (RM) a l’hora de fer prediccions.

In [31]:
X_no_rm = housing_ML.drop(['RM','MEDV'], axis=1) 
y_no_rm = np.array(housing_ML['MEDV']) 

In [33]:
X_train_no_rm, X_test_no_rm, y_train_no_rm, y_test_no_rm = train_test_split(X_no_rm, y_no_rm, test_size=0.2, random_state=42)

print('Forma del Training set: X={}, y={}'.format(X_train_no_rm.shape, y_train_no_rm.shape))
print('Forma del Test set: X={}, y={}'.format(X_test_no_rm.shape, y_test_no_rm.shape))

Forma del Training set: X=(404, 20), y=(404,)
Forma del Test set: X=(102, 20), y=(102,)


In [35]:
#Decision Tree Regressor
dt_no_rm = DecisionTreeRegressor(random_state=42)
dt_no_rm.fit(X_train_no_rm, y_train_no_rm)

DecisionTreeRegressor(random_state=42)

In [36]:
predictions_dt_no_rm = dt_no_rm.predict(X_test_no_rm)

In [37]:
#Random Forest
rf_no_rm = RandomForestRegressor(random_state=42)
rf_no_rm.fit(X_train_no_rm, y_train_no_rm)

RandomForestRegressor(random_state=42)

In [38]:
predictions_rf_no_rm = rf_no_rm.predict(X_test_no_rm)

In [39]:
#Avaluació dels models
print('MSE del Decision Tree Regression:', mean_squared_error(y_test_no_rm, predictions_dt_no_rm))
print('R2 del Decision Tree Regression:', r2_score(y_test_no_rm, predictions_dt_no_rm))

print('MSE del Random Forest:', mean_squared_error(y_test_no_rm, predictions_rf_no_rm))
print('R2 del Random Forest:', r2_score(y_test_no_rm, predictions_rf_no_rm))

MSE del Decision Tree Regression: 25.577352941176468
R2 del Decision Tree Regression: 0.6512200538562283
MSE del Random Forest: 13.654145509803916
R2 del Random Forest: 0.8138082487855139


Començant pel MSE, els valors obtinguts anteriorment eren:
- MSE del Decision Tree Regression: 11.331568627450983
- MSE del Random Forest: 8.012439598039217.

Per tant, veiem que quan eliminem la variable del nombre d'habitacions, RM, obtenim un major error i, per tant, un pitjor rendiment.

El mateix succeeix quan analitzem el valor R2. En aquest cas els resultats obtinguts eren:
- R2 del Decision Tree Regression: 0.8454795574547436
- R2 del Random Forest: 0.8907401302272603

Per tant, quan eliminem la variable RM el R2 disminueix, el que ens indica que el model és menys capaç d'explicar la variabilitat de la varible objectiu (el valor mitjà dels inmobles).

Això vol dir que el nombre d'habitacions constitueix una variable important pel model, que convé tenir en compte per a poder millorar les prediccions.